### PR TITLE
Minor fix in optimizer + docs

### LINKED ISSUE #1259: [BUG]
**Describe the bug**
I am trying to set up my code so that it periodically saves the model and optimizer states (with a predefined checkpoint frequency) so that if desired I can go back to reload the model and optimizer states and continue the training from any of the saved checkpoints. I am using `@compile` with the train_step function. If I simply save and reload without actually quitting the code the training seems to continue as expected based on the loss decrease trend. However, if I actually quit the code and try to reload in a fresh run there seems to be jump in loss, so it seems that I am not exactly reproducing the state of the training. Below are relevant code snippets.   

**To Reproduce**

Include code snippet
```python
def save_model_and_optimizer(config, current_epoch, dir_path, model_base_file_name, optimizer_base_file_name, hyper_base_file_name):
    if not os.path.exists(dir_path):
        os.makedirs(dir_path)

    model_file_name = f"{model_base_file_name}_epoch_{current_epoch}.pkl"
    optimizer_file_name = f"{optimizer_base_file_name}_epoch_{current_epoch}.pkl"
    random_state_file_name = f"random_state_epoch_{current_epoch}.pkl"
    config_file_name = f"{hyper_base_file_name}_epoch_{current_epoch}.json"

    # Save model state (parameters only)
    model_file_path = os.path.join(dir_path, model_file_name)
    with open(model_file_path, 'wb') as f:
        pickle.dump(model.parameters(), f)

    # Save optimizer state
    optimizer_file_path = os.path.join(dir_path, optimizer_file_name)
    with open(optimizer_file_path, 'wb') as f:
        pickle.dump(optimizer.state, f)

    # Save random state
    random_state_path = os.path.join(dir_path, random_state_file_name)
    with open(random_state_path, 'wb') as f:
        pickle.dump(mx.random.state, f)

    # Save training configuration
    config['current_epoch'] = current_epoch
    hyper_file_path = os.path.join(dir_path, config_file_name)
    with open(hyper_file_path, 'w') as json_file:
        json.dump(config, json_file, indent=4)

    print(f"Model, optimizer, random state, and configuration saved at epoch {current_epoch}.")

def load_model_and_optimizer(model, optimizer, dir_path, model_base_file_name, optimizer_base_file_name, hyper_base_file_name, checkpoint_epoch):
    model_file_name = f"{model_base_file_name}_epoch_{checkpoint_epoch}.pkl"
    optimizer_file_name = f"{optimizer_base_file_name}_epoch_{checkpoint_epoch}.pkl"
    random_state_file_name = f"random_state_epoch_{checkpoint_epoch}.pkl"
    config_file_name = f"{hyper_base_file_name}_epoch_{checkpoint_epoch}.json"

    model_file_path = os.path.join(dir_path, model_file_name)
    optimizer_file_path = os.path.join(dir_path, optimizer_file_name)
    random_state_file_path = os.path.join(dir_path, random_state_file_name)
    config_file_path = os.path.join(dir_path, config_file_name)

    if os.path.exists(model_file_path) and os.path.exists(optimizer_file_path) and os.path.exists(random_state_file_path) and os.path.exists(config_file_path):
        # Load model state (parameters only)
        with open(model_file_path, 'rb') as f:
            parameters = pickle.load(f)
            model.update(parameters)

        # Load optimizer state
        with open(optimizer_file_path, 'rb') as f:
            optimizer.state = pickle.load(f)

        # Load random state
        with open(random_state_file_path, 'rb') as f:
            mx.random.state = pickle.load(f)

        # Load training configuration
        with open(config_file_path, 'r') as json_file:
            config = json.load(json_file)

        start_epoch = config.get('current_epoch', 0)
        print(f"Model, optimizer, random state, and configuration loaded from {dir_path} at epoch {checkpoint_epoch}.")
    else:
        start_epoch = 0
        print(f"No saved model found at epoch {checkpoint_epoch}. Starting training from scratch.")

    return start_epoch, config


    @partial(mx.compile, inputs=state, outputs=state)
    def train_step(src, target, src_alphas, src_dts, dx, dy):
        loss_and_grad_fn = nn.value_and_grad(model, loss_fn_2D)
        loss, grads = loss_and_grad_fn(model, src, target, src_alphas, src_dts, dx, dy)
        optimizer.update(model, grads)
        return loss, grads

    @partial(mx.compile, inputs=state, outputs=state)
    def evaluate_step(src, target, src_alphas, src_dts, dx, dy):
        loss = loss_fn_2D(model, src, target, src_alphas, src_dts, dx, dy)
        return loss
```

**Expected behavior**
During the initial run the loss history is as follows 

```
start_epoch: 0 epochs: 10 learing_rate: array(0, dtype=float32)
Epoch 1, Training Loss: 12.25573897861299, Validation Loss: 12.527009956240654
Epoch 2, Training Loss: 12.255739007620585, Validation Loss: 12.527009927630424
Model, optimizer, random state, and configuration saved at epoch 2.
Epoch 3, Training Loss: 0.250323773617191, Validation Loss: 0.09001006038859487
Epoch 4, Training Loss: 0.05461768370742599, Validation Loss: 0.03122760124504566
Model, optimizer, random state, and configuration saved at epoch 4.
Epoch 5, Training Loss: 0.03391449196275235, Validation Loss: 0.014082754662958906
```

I try to reload from the saved state at epoch 4 and expect after one more epoch of training to get a loss in the range of Epoch 5 in the initial training history. Instead I get 

```
Model, optimizer, random state, and configuration loaded from /Users/m2/PycharmProjects/pythonProject_StreamLit/Transformer_save at epoch 4.
************************ Hostname: PaloAlto Starting Training *********************** 
start_epoch: 4 epochs: 10 learing_rate: array(0.0001, dtype=float32)
Epoch 5, Training Loss: 0.31006237406904497, Validation Loss: 0.06310294904746115
```

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.1.2]
 - Version [e.g. 0.7.0]

**Additional context**

### PR DESCRIPTION
**Pull request by awni**
Previously if you set the state of an uninitialized optimizer then it would silently reinitialize it on the first update.

This changes that behavior so that you can set the state without needing to worry about making sure the optimizer is initialized.

Also added a doc on saving and loading optimizer state.

Closes #1259

### CODE REVIEW

**Review by angeloskath — APPROVED** [review]
🙏
